File: Android/app/src/main/java/realm/io/realmpop/model/Game.java
package realm.io.realmpop.model;

import java.util.Arrays;

import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.annotations.Required;
import realm.io.realmpop.util.PopUtils;

import static realm.io.realmpop.util.RandomNumberUtils.generateNumbersArray;

public class Game extends RealmObject {

    private Side player1;
    private Side player2;

    @Required
    private String numbers;

    public Game() {}

    public Game(Player me, Player challenger, int[] numberArray) {
        player1 = new Side(me, numberArray.length);
        player2 = new Side(challenger, numberArray.length);
        numbers = numberArrayToString(numberArray);
    }

    public Side getPlayer1() { return player1; }

    public void setPlayer1(Side player1) { this.player1 = player1; } 

    public Side getPlayer2() { return player2; }

    public void setPlayer2(Side player2) { this.player2 = player2; } 

    public String getNumbers() { return numbers; }

    public void setNumbers(String numbers) { this.numbers = numbers; } 

    public int[] getNumberArray() {
        if(getNumbers() == null) {
            return new int[]{};
        } else {
            String [] numberStrArr = getNumbers().split(",");
            int[] numArr = new int[numberStrArr.length];
            for(int i = 0; i < numArr.length; i++) {
                numArr[i] = Integer.valueOf(numberStrArr[i]);
            }
            return numArr;
        }
    }

    public boolean isGameOver() {
        return player1.isFailed() || player2.isFailed();
    }

    public Side sideWithPlayerId(final String playerId) {
        if(player1 != null && player1.getPlayerId().equals(playerId)) {
            return player1;
        } else if (player2 != null && player2.getPlayerId().equals(playerId)) {
            return player2;
        } else {
            return null;
        }
    }

    private static String numberArrayToString(int[] numArray) {
        String numStr = Arrays.toString(numArray);
        return numStr.replaceAll("\\[|\\]|\\s", "");
    }

}

File: Android/app/src/main/java/realm/io/realmpop/model/Game.kt
package realm.io.realmpop.model

import io.realm.RealmObject
import io.realm.annotations.Required
import java.util.*

open class Game : RealmObject {

    var player1: Side? = null
    var player2: Side? = null
    @Required
    var numbers: String? = null

    constructor() {}
    constructor(me: Player, challenger: Player, numberArray: IntArray) {
        player1 = Side(me, numberArray.size)
        player2 = Side(challenger, numberArray.size)
        numbers = numberArrayToString(numberArray)
    }

    val numberArray: IntArray
        get() {
            if (numbers == null) {
                return intArrayOf()
            } else {
                val numberStrArr = numbers!!
                        .split(",".toRegex())
                        .dropLastWhile { it.isEmpty() }
                        .toTypedArray()
                val numArr = IntArray(numberStrArr.size)
                for (i in numArr.indices) {
                    numArr[i] = Integer.valueOf(numberStrArr[i])!!
                }
                return numArr
            }
        }

    val isGameOver: Boolean
        get() = player1!!.failed || player2!!.failed


    fun sideWithPlayerId(playerId: String) : Side? {
        return when {
            player1?.playerId.equals(playerId) -> player1
            player2?.playerId.equals(playerId) -> player2
            else -> null
        }
    }

    private fun numberArrayToString(numArray: IntArray): String {
        val numStr = Arrays.toString(numArray)
        return numStr.replace("\\[|\\]|\\s".toRegex(), "")
    }

}
